Sources/Scheduler.swift

```diff
@@ -382,6 +382,90 @@ public final class QueueScheduler: DateScheduler {
 	}
 }
 
+/// A scheduler backed by a run loop.
+public final class RunLoopScheduler: DateScheduler {
+	/// A singleton `RunLoopScheduler` that always targets the main run loop.
+	static let main = RunLoopScheduler(runLoop: .main)
+
+	private let runLoop: CFRunLoop
+
+	/// Initializes a scheduler that will target the provided run loop in a thread
+	/// safe way.
+	///
+	/// - parameters:
+	///   - runLoop: The run loop that will be used to schedule work.
+	///
+	/// - note: Like `UIScheduler`, this scheduler will execute work synchronously
+	///         whenever possible, if it detects that it is already running under
+	///         the specified run loop.
+	///
+	/// - note: The scheduler takes no responsibility for running the run loop.
+	///         It assumes that the run loop is already running and configured
+	///         correctly.
+	public init(runLoop: RunLoop) {
+		self.runLoop = runLoop.getCFRunLoop()
+	}
+
+	public var currentDate: Date {
+		return Date()
+	}
+
+	public func schedule(_ action: @escaping () -> Void) -> Disposable? {
+		if CFRunLoopGetCurrent() == runLoop {
```

Comment:
FIFO order needs to be maintained in case of recursive scheduling.

Reply:
I’ve been thinking about this more today, and I think that because of `CFRunLoopRun()` and friends, we’ll have to use a different technique, because callers can always recursively run the run loop out of our control.

I think we might be able to maintain some thread-local state that tracks whether we’re already running inside `RunLoopScheduler`, and then potentially reschedule if that’s already the case?

But again, the problem is this call tree:

```
RunLoopScheduler.schedule(_:)         // (1) sets thread-local flag
    CFRunLoopRun()                    // (2) potentially spins indefinitely
        RunLoopScheduler.schedule(_:) // (3) we'll never be able to schedule unless we call CFRunLoopStop(_:)?
```

Another alternative may be to simply not provide any guarantees: if you’re opting into run loop scheduling, be aware of the pitfalls of recursively spinning the run loop.